# Solution to exercise 1 

In [1]:
# Importing necessary libraries
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import f1_score

#Use 42 as Random_Variable to produce the same results
RANDOM_VARIABLE = 42

In [2]:
# Loading the dataset and split it o train and test set. Printing the size of sets and the unique classes of the dataset
X,y = load_breast_cancer(return_X_y=True)
X_train,X_test,y_train,y_test = train_test_split(X, y, test_size=0.30, random_state=RANDOM_VARIABLE)
print("Size of train set:{}".format(len(y_train)))
print("Size of test set:{}".format(len(y_test)))
print("Unique classes:{}".format(len(set(y_test))))

Size of train set:398
Size of test set:171
Unique classes:2


<p><em>Creating two Decision Tree Classifiers. The one uses gini as criterion and the other entropy. Fit classifiers with the train data, store the predictions to prediction_gini and prediction_igain. Finally calculate the f1_score and print it for both of classifiers.</em></p>

In [3]:
classifier_gini = DecisionTreeClassifier(criterion="gini", random_state=RANDOM_VARIABLE)
classifier_igain = DecisionTreeClassifier(criterion="entropy", random_state=RANDOM_VARIABLE)

classifier_gini.fit(X_train,y_train)
classifier_igain.fit(X_train,y_train)

prediction_gini = classifier_gini.predict(X_test)
prediction_igain = classifier_igain.predict(X_test)

f_measure_gini = f1_score(y_test, prediction_gini)
f_measure_igain = f1_score(y_test, prediction_igain)
print("F-Measure Gini:{}".format(f_measure_gini))
print("F-Measure Information Gain:{}".format(f_measure_igain))

F-Measure Gini:0.9528301886792453
F-Measure Information Gain:0.9724770642201834


<p><em>Creating a loop from 1 to max_depth+1 from classifier_gini and store the f1_scores the DecisionTreeClassifier achieves. We create a new DecisionTreeClassifier (with criterion='gini', max_depth=i) at every single loop, fit the data, predict (in the first loop for the train set, in the second for the test set) and store the f1_score. Printing the lists with f1_scores  </em></p>

In [5]:
depth = classifier_gini.tree_.max_depth
fscores_train = []
fscores_test = []

for i in range(1,depth+1):
    classifier_gini = DecisionTreeClassifier(criterion="gini", max_depth=i,random_state=RANDOM_VARIABLE)
    classifier_gini.fit(X_train,y_train)
    prediction_gini = classifier_gini.predict(X_train)
    f_measure_gini = f1_score(y_train, prediction_gini)
    fscores_train.append(f_measure_gini)
    
for i in range(1,depth+1):
    classifier_gini = DecisionTreeClassifier(criterion="gini", max_depth=i,random_state=RANDOM_VARIABLE)
    classifier_gini.fit(X_train,y_train)
    prediction_gini = classifier_gini.predict(X_test)
    f_measure_gini = f1_score(y_test, prediction_gini)
    fscores_test.append(f_measure_gini)


print("Fscores Train:{}".format(fscores_train))
print("Fscores Test:{}".format(fscores_test))

Fscores Train:[0.9392712550607287, 0.9533468559837729, 0.9761904761904762, 0.996, 0.996, 0.9979959919839679, 1.0]
Fscores Test:[0.9150943396226415, 0.9444444444444444, 0.9724770642201834, 0.9629629629629629, 0.9629629629629629, 0.9674418604651163, 0.9528301886792453]


<b>Compare the results from the train set with the results from the test set. What do you notice? Explain your findings. How are you going to choose the max_depth of your model?</b>

<p><em>It is observed that as the depth of the tree increases in the case of the train set, the accuracy increases until it reaches 100%. However, the accuracy of the test data increases to depth 3 where it gets a maximum value of 97.2% and from there on it starts to fall. This is due to the fact that the tree is starting to become more restrictive than it should be and is starting to overfit the data. This is shown by the fact that the train score increases while the test score decreases. The best depth maximum depth for this tree is 3.</em><p>